posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,133 @@
+from constance import config
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.events import EVENTS_TABLE, EVENTS_TABLE_SQL
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.version_requirement import ServiceVersionRequirement
+
+TEMPORARY_TABLE_NAME = "temp_events_0001_events_sample_by"
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Events table migration for compatible sample by column."
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=EVENTS_TABLE_SQL().replace(EVENTS_TABLE, TEMPORARY_TABLE_NAME, 1),
+            rollback=f"DROP TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp < toYYYYMM(now()) - 1
+            AND timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
+            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"DETACH TABLE {EVENTS_TABLE}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"ATTACH TABLE {EVENTS_TABLE}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
```

Comment:
Wouldn't we be in a bad state, if say a future step fails and rollback fails too and we never get to here?

Reply:
Get where? I'm sure there are plenty of ways this can fail into fun inconsistent states. We have to just be defensive about the ones that are most likely to happen 🛡️

Reply:
Specifically in the next 2 steps we might want them to have `side_effect_rollback=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", True)`, so we will start ingestion back up sooner